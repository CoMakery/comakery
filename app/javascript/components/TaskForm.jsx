import React from 'react'
import PropTypes from 'prop-types'
import ProjectSetup from './layouts/ProjectSetup'
import {fetch as fetchPolyfill} from 'whatwg-fetch'
import InputFieldWhiteDark from './styleguide/InputFieldWhiteDark'
import InputFieldDescription from './styleguide/InputFieldDescription'
import InputFieldDescriptionLarge from './styleguide/InputFieldDescriptionLarge'
import InputFieldUploadFile from './styleguide/InputFieldUploadFile'
import InputFieldDropdown from './styleguide/InputFieldDropdown'
import Button from './styleguide/Button'
import ButtonBorder from './styleguide/ButtonBorder'
import Flash from './layouts/Flash'
import SidebarItem from './styleguide/SidebarItem'

class TaskForm extends React.Component {
  constructor(props) {
    super(props)

    this.errorAdd = this.errorAdd.bind(this)
    this.errorRemove = this.errorRemove.bind(this)
    this.disable = this.disable.bind(this)
    this.enable = this.enable.bind(this)
    this.handleFieldChange = this.handleFieldChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    this.state = {
      flashMessages                         : [],
      errors                                : {},
      disabled                              : {},
      formAction                            : this.props.formAction,
      formUrl                               : this.props.formUrl,
      closeOnSuccess                        : false,
      'task[name]'                          : this.props.task.name || '',
      'task[why]'                           : this.props.task.why || '',
      'task[description]'                   : this.props.task.description || '',
      'task[requirements]'                  : this.props.task.requirements || '',
      'task[experience_level]'              : this.props.task.experienceLevel || '',
      'task[amount]'                        : this.props.task.amount || '',
      'task[number_of_assignments]'         : this.props.task.numberOfAssignments || 1,
      'task[number_of_assignments_per_user]': this.props.task.numberOfAssignmentsPerUser || 1,
      'task[expires_in_days]'               : this.props.task.expiresInDays || 10,
      'task[specialty_id]'                  : this.props.task.specialtyId || Object.values(this.props.specialties)[0],
      'task[proof_link]'                    : this.props.task.proofLink || ''
    }
  }

  goBack() {
    typeof window === 'undefined' ? null : window.location = document.referrer
  }

  errorAdd(n, e) {
    this.setState({
      errors: Object.assign({}, this.state.errors, {[n]: e})
    })
  }

  errorRemove(n) {
    let e = this.state.errors
    delete e[n]
    this.setState({
      errors: e
    })
  }

  disable(a) {
    let d = this.state.disabled
    a.forEach(n => d = Object.assign({}, d, {[n]: true}))
    this.setState({
      disabled: d
    })
  }

  enable(a) {
    let d = this.state.disabled
    a.forEach(n => delete d[n])
    this.setState({
      disabled: d
    })
  }

  handleFieldChange(event) {
    this.setState({ [event.target.name]: event.target.value })

    if (event.target.name === 'task[number_of_assignments]' && event.target.value === '1') {
      this.setState({ 'task[number_of_assignments_per_user]': '1' })
    }

    if (!event.target.checkValidity()) {
      this.errorAdd(event.target.name, 'invalid value')
      return
    } else {
      this.errorRemove(event.target.name)
    }

    if (event.target.value === '') {
      return
    }
  }

  handleSubmit(event) {
    event.preventDefault()

    this.disable(['task[submit]', 'task[submit_and_close]'])

    if (!event.target.checkValidity()) {
      this.enable(['task[submit]', 'task[submit_and_close]'])
      return
    }

    const formData = new FormData(event.target)

    fetchPolyfill(this.state.formUrl, {
      credentials: 'same-origin',
      method     : this.state.formAction,
      body       : formData,
      headers    : {
        'X-Key-Inflection': 'snake'
      }
    }).then(response => {
      if (response.status === 200) {
        if (this.state.closeOnSuccess) {
          window.location = this.props.urlOnSuccess
        } else {
          response.json().then(data => {
            if (this.state.formAction === 'POST') {
              this.setState(state => ({
                formAction   : 'PUT',
                formUrl      : data.formUrl,
                flashMessages: state.flashMessages.concat([{'severity': 'notice', 'text': data.message}])
              }))
              history.replaceState(
                {},
                document.title,
                `${window.location.origin}${data.editUrl}`
              )
            } else {
              this.setState(state => ({
                flashMessages: state.flashMessages.concat([{'severity': 'notice', 'text': data.message}])
              }))
            }
          })
          this.enable(['task[submit]', 'task[submit_and_close]'])
        }
      } else {
        response.json().then(data => {
          this.setState(state => ({
            errors       : data.errors,
            flashMessages: state.flashMessages.concat([{'severity': 'error', 'text': data.message}])
          }))
          this.enable(['task[submit]', 'task[submit_and_close]'])
        })
      }
    })
  }

  render() {
    return (
      <React.Fragment>
        <ProjectSetup
          className="task-form"
          projectForHeader={this.props.projectForHeader}
          missionForHeader={this.props.missionForHeader}
          owner
          current="batches"
          hasBackButton
          subfooter={
            <React.Fragment>
              <Button
                value={this.state.formAction === 'POST' ? 'create & close' : 'save & close'}
                type="submit"
                form="task-form--form"
                disabled={this.state.disabled['task[submit_and_close]']}
                onClick={() => this.setState({closeOnSuccess: true})}
              />
              <ButtonBorder
                value={this.state.formAction === 'POST' ? 'create' : 'save'}
                type="submit"
                form="task-form--form"
                disabled={this.state.disabled['task[submit]']}
                onClick={() => this.setState({closeOnSuccess: false})}
              />
              <ButtonBorder
                value="cancel"
                onClick={this.goBack}
              />
            </React.Fragment>
          }
          sidebar={
            <React.Fragment>
              <div className="batch-index--sidebar">
                <SidebarItem
                  className="batch-index--sidebar--item batch-index--sidebar--item__form"
                  text={this.props.batch.name}
                  selected
                />
                <hr className="batch-index--sidebar--hr" />
              </div>
            </React.Fragment>
          }
        >
          <Flash messages={this.state.flashMessages} />

          <div className="task-form--form--title">
            {this.state.formAction === 'POST' ? 'Create a New Task' : 'Edit Task'}
          </div>

          <form className="task-form--form" id="task-form--form" onSubmit={this.handleSubmit}>
            <InputFieldWhiteDark
              title="name"
              required
              name="task[name]"
              value={this.state['task[name]']}
              errorText={this.state.errors['task[name]']}
              eventHandler={this.handleFieldChange}
              placeholder="Provide a clear title for the Task"
              symbolLimit={100}
            />

            <InputFieldDescription
              title="WHAT IS THE EXPECTED BENEFIT"
              required
              name="task[why]"
              value={this.state['task[why]']}
              errorText={this.state.errors['task[why]']}
              eventHandler={this.handleFieldChange}
              placeholder="Use the following format to describe the expected benefit of this task being completed: As a ___, I want to ___,  So I can  ___. Example: As a new user, I want to be able to sign up, so I can claim my 10% discount."
              symbolLimit={500}
            />

            <InputFieldDescriptionLarge
              title="acceptance requirements"
              required
              name="task[requirements]"
              value={this.state['task[requirements]']}
              errorText={this.state.errors['task[requirements]']}
              eventHandler={this.handleFieldChange}
              placeholder="Create bullet points for each acceptance criteria by starting a line with an asterisk (markdown format). These bullet points will be used by reviewers to verify the work."
              symbolLimit={1000}
            />

            <InputFieldDescriptionLarge
              title="description"
              name="task[description]"
              value={this.state['task[description]']}
              errorText={this.state.errors['task[description]']}
              eventHandler={this.handleFieldChange}
              placeholder="Provide a longer description about the task, the type of work involved, and how it will relate to the larger batch"
              symbolLimit={500}
            />

            <InputFieldUploadFile
              title="image asset to help guide the task"
              name="task[image]"
              errorText={this.state.errors['task[image]']}
              imgPreviewUrl={this.props.task.imageUrl}
              imgPreviewDimensions="100x100"
            />

            {this.props.task.imageFromId &&
              <input
                type="hidden"
                name="task[image_from_id]"
                value={this.props.task.imageFromId}
                readOnly
              />
            }

            <InputFieldDropdown
              title="specialty"
              required
              name="task[specialty_id]"
              value={this.state['task[specialty_id]']}
              errorText={this.state.errors['task[specialtyId]']}
              eventHandler={this.handleFieldChange}
              selectEntries={Object.entries(this.props.specialties)}
              symbolLimit={0}
            />

            <InputFieldDropdown
              title="skill confirmations required"
              required
              name="task[experience_level]"
              value={this.state['task[experience_level]']}
              errorText={this.state.errors['task[experienceLevel]']}
              eventHandler={this.handleFieldChange}
              selectEntries={Object.entries(this.props.experienceLevels)}
              symbolLimit={0}
            />

            <InputFieldWhiteDark
              title={`award amount (${this.props.token.symbol || 'no token'})`}
              required
              name="task[amount]"
              value={this.state['task[amount]']}
              errorText={this.state.errors['task[amount]']}
              eventHandler={this.handleFieldChange}
              type="number"
              min="0"
              step={`${1.0 / (10 ** this.props.token.decimalPlaces)}`}
              placeholder="The total amount of tokens or coins you are paying for this task to be completed"
              symbolLimit={0}
            />

            <InputFieldWhiteDark
              title="how many times can this task be done"
              required
              name="task[number_of_assignments]"
              value={this.state['task[number_of_assignments]']}
              errorText={this.state.errors['task[numberOfAssignments]']}
              eventHandler={this.handleFieldChange}
              type="number"
              min="1"
              step="1"
              placeholder="1"
              symbolLimit={0}
            />

            <InputFieldWhiteDark
              title="how many times can this task be done by a single user"
              required
              name="task[number_of_assignments_per_user]"
              value={this.state['task[number_of_assignments_per_user]']}
              errorText={this.state.errors['task[numberOfAssignmentsPerUser]']}
              eventHandler={this.handleFieldChange}
              type="number"
              readOnly={this.state['task[number_of_assignments]'] === '1'}
              min="1"
              step="1"
              placeholder="1"
              symbolLimit={0}
            />

            <InputFieldWhiteDark
              title="days till task expires (after starting)"
              required
              name="task[expires_in_days]"
              value={this.state['task[expires_in_days]']}
              errorText={this.state.errors['task[expiresInDays]']}
              eventHandler={this.handleFieldChange}
              type="number"
              min="1"
              step="1"
              placeholder="10"
              symbolLimit={0}
            />

            <InputFieldWhiteDark
              title="URL where to submit completed work"
              name="task[proof_link]"
              value={this.state['task[proof_link]']}
              errorText={this.state.errors['task[proofLink]']}
              eventHandler={this.handleFieldChange}
              placeholder="URL to a Dropbox or Drive Folder, or a GitHub Repo (http://example.com)"
              symbolLimit={150}
            />

            <input
              type="hidden"
              name="authenticity_token"
              value={this.props.csrfToken}
              readOnly
            />
          </form>
        </ProjectSetup>
      </React.Fragment>
    )
  }
}

TaskForm.propTypes = {
  task            : PropTypes.object.isRequired,
  batch           : PropTypes.object.isRequired,
  project         : PropTypes.object.isRequired,
  token           : PropTypes.object.isRequired,
  experienceLevels: PropTypes.object.isRequired,
  specialties     : PropTypes.object.isRequired,
  formUrl         : PropTypes.string.isRequired,
  formAction      : PropTypes.string.isRequired,
  urlOnSuccess    : PropTypes.string.isRequired,
  csrfToken       : PropTypes.string.isRequired,
  missionForHeader: PropTypes.object,
  projectForHeader: PropTypes.object
}
TaskForm.defaultProps = {
  task            : {'default': '_'},
  batch           : {'default': '_'},
  project         : {'default': '_'},
  token           : {'default': '_'},
  experienceLevels: {'default': '_'},
  specialties     : {'default': '_'},
  formUrl         : '/',
  formAction      : 'POST',
  urlOnSuccess    : '/',
  csrfToken       : '00',
  missionForHeader: null,
  projectForHeader: null
}
export default TaskForm
