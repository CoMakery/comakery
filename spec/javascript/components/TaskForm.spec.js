import React from 'react'
import { mount } from 'enzyme'
import TaskForm from 'components/TaskForm'

describe('TaskForm', () => {
  it('renders correctly without props', () => {
    const wrapper = mount(<TaskForm />)

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.exists('Layout[title]')).toBe(true)
    expect(wrapper.exists('.task-form')).toBe(true)
    expect(wrapper.exists('ButtonBorder[value="cancel"]')).toBe(true)
    expect(wrapper.exists('ButtonBorder[value="create"]')).toBe(true)
    expect(wrapper.exists('Button[value="create & close"]')).toBe(true)
    expect(wrapper.exists('.task-form--message')).toBe(false)
    expect(wrapper.exists('.task-form--form')).toBe(true)
    expect(wrapper.exists('#task-form--form')).toBe(true)

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="name"][required][name="task[name]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDescriptionMiddle[title="why"][required][name="task[why]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDescriptionMiddle[title="description"][required][name="task[description]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDescription[title="acceptance requirements"][required][name="task[requirements]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDropdown[title="specialty"][required][name="task[specialty_id]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="award amount (no token)"][required][name="task[amount]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldWhiteDark[title="URL where to submit completed work"][required][name="task[proof_link]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'input[type="hidden"][name="authenticity_token"]'
    )).toBe(true)
  })

  it('renders correctly with specialties', () => {
    const specialties = {
      '1': 1,
      '2': 2
    }
    const wrapper = mount(<TaskForm specialties={specialties} />)

    expect(wrapper.find(
      'InputFieldDropdown[title="specialty"][required][name="task[specialty_id]"]'
    ).props().value).toBe(1)

    expect(wrapper.find(
      'InputFieldDropdown[title="specialty"][required][name="task[specialty_id]"]'
    ).props().selectEntries).toEqual(Object.entries(specialties))
  })

  it('renders correctly with a task and token', () => {
    const task = {
      'id'          : 28,
      'description' : 'Task description',
      'awardTypeId' : 10,
      'proofLink'   : 'http://nowhere',
      'name'        : 'Task name',
      'why'         : 'Task why',
      'requirements': 'Task requirements',
      'amount'      : '100.0',
    }
    const token = {
      'symbol'       : 'DMT',
      'decimalPlaces': 8,
    }

    const wrapper = mount(<TaskForm
      task={task}
      token={token}
    />)

    expect(wrapper.find(
      'InputFieldWhiteDark[title="name"][required][name="task[name]"]'
    ).props().value).toBe('Task name')

    expect(wrapper.find(
      'InputFieldDescriptionMiddle[title="why"][required][name="task[why]"]'
    ).props().value).toBe('Task why')

    expect(wrapper.find(
      'InputFieldDescriptionMiddle[title="description"][required][name="task[description]"]'
    ).props().value).toBe('Task description')

    expect(wrapper.find(
      'InputFieldDescription[title="acceptance requirements"][required][name="task[requirements]"]'
    ).props().value).toBe('Task requirements')

    expect(wrapper.find(
      'InputFieldWhiteDark[title="award amount (DMT)"][required][name="task[amount]"]'
    ).props().value).toBe('100.0')

    expect(wrapper.find(
      'InputFieldWhiteDark[title="URL where to submit completed work"][required][name="task[proof_link]"]'
    ).props().value).toBe('http://nowhere')
  })

  it('renders correctly with csrfToken', () => {
    const wrapper = mount(<TaskForm csrfToken="test" />)

    expect(wrapper.find(
      'input[type="hidden"][name="authenticity_token"]'
    ).props().value).toBe('test')
  })

  it('uses formUrl', () => {
    const wrapper = mount(<TaskForm formUrl="/test" />)

    expect(wrapper.state('formUrl')).toBe('/test')
  })

  it('uses formAction', () => {
    const wrapper = mount(<TaskForm formAction="PUT" />)

    expect(wrapper.state('formAction')).toBe('PUT')
  })

  it('uses urlOnSuccess', () => {
    const wrapper = mount(<TaskForm urlOnSuccess="/test" />)

    expect(wrapper.props().urlOnSuccess).toBe('/test')
  })

  it('displays messages', () => {
    const wrapper = mount(<TaskForm />)

    wrapper.setState({
      flashMessages: [
        {
          'severity': 'notice',
          'text'    : 'notice text'
        },
        {
          'severity': 'warning',
          'text'    : 'warning text'
        },
        {
          'severity': 'error',
          'text'    : 'error text'
        },
      ]
    })

    wrapper.update()

    expect(wrapper.exists('Flash')).toBe(true)
  })

  it('displays errors', () => {
    const wrapper = mount(<TaskForm />)

    wrapper.setState({
      errors: {
        'task[name]'       : 'name error',
        'task[description]': 'description error',
      }
    })

    wrapper.update()

    expect(wrapper.exists(
      'InputFieldWhiteDark[errorText="name error"][title="name"][required][name="task[name]"]'
    )).toBe(true)

    expect(wrapper.exists(
      'InputFieldDescriptionMiddle[errorText="description error"][title="description"][required][name="task[description]"]'
    )).toBe(true)
  })
})
