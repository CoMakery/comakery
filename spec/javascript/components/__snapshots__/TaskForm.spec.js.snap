// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm renders correctly without props 1`] = `
<TaskForm
  csrfToken="00"
  formAction="POST"
  formUrl="/"
  task={
    Object {
      "default": "_",
    }
  }
  token={
    Object {
      "default": "_",
    }
  }
  urlOnSuccess="/"
>
  <Layout
    category=""
    className="task-form"
    hasBackButton={true}
    navTitle={null}
    sidebar={null}
    subfooter={
      <React.Fragment>
        <Button
          className=""
          disabled={false}
          form="task-form--form"
          onClick={[Function]}
          type="submit"
          value="create & close"
        />
        <ButtonBorder
          className=""
          form="task-form--form"
          onClick={[Function]}
          type="submit"
          value="create"
        />
        <ButtonBorder
          className=""
          onClick={[Function]}
          value="cancel"
        />
      </React.Fragment>
    }
    title="Create a New Task"
  >
    <div
      className="layout task-form"
    >
      <div
        className="layout--back-button"
      >
        <Icon
          className=""
          name="iconCloseCopy.svg"
          onClick={[Function]}
        >
          <img
            className="icon icon__iconCloseCopy-svg"
            onClick={[Function]}
            src="test-file-stub"
          />
        </Icon>
      </div>
      <div
        className="layout--content"
      >
        <div
          className="layout--content--title"
        >
          Create a New Task
        </div>
        <hr
          className="layout--content--hr"
        />
        <div
          className="layout--content--wrapper"
        >
          <div
            className="layout--content--content"
          >
            <Flash
              className=""
              messages={Array []}
            >
              <div
                className="flash-message-container"
              />
            </Flash>
            <form
              className="task-form--form"
              id="task-form--form"
              onSubmit={[Function]}
            >
              <InputFieldWhiteDark
                eventHandler={[Function]}
                name="task[name]"
                placeholder="Provide a clear title for the Task"
                required={true}
                symbolLimit={100}
                title="name"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__white__dark"
                  copyOnClick={false}
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgInputRef={null}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  imgRequirements=""
                  imgVerifier={[Function]}
                  max=""
                  min=""
                  name="task[name]"
                  pattern=".*"
                  placeholder="Provide a clear title for the Task"
                  readOnly={false}
                  recommended={false}
                  required={true}
                  selectEntries={Array []}
                  step=""
                  symbolLimit={100}
                  title="name"
                  type="text"
                  value=""
                >
                  <div
                    className="input-field input-field__white__dark"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        name
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                      <span
                        className="input-field--title--counter"
                      >
                        0
                        /
                        100
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <input
                        className="input-field--content__text"
                        name="task[name]"
                        onChange={[Function]}
                        onClick={null}
                        pattern=".*"
                        placeholder="Provide a clear title for the Task"
                        readOnly={false}
                        required={true}
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldWhiteDark>
              <InputFieldDescriptionMiddle
                eventHandler={[Function]}
                name="task[why]"
                placeholder="Let people know what will be the result of this task being completed (ex: When task is finished, the result will be xxxx)"
                required={true}
                symbolLimit={500}
                title="why"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__description-middle"
                  copyOnClick={false}
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgInputRef={null}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  imgRequirements=""
                  imgVerifier={[Function]}
                  max=""
                  min=""
                  name="task[why]"
                  pattern=".*"
                  placeholder="Let people know what will be the result of this task being completed (ex: When task is finished, the result will be xxxx)"
                  readOnly={false}
                  recommended={false}
                  required={true}
                  selectEntries={Array []}
                  step=""
                  symbolLimit={500}
                  title="why"
                  type="textarea"
                  value=""
                >
                  <div
                    className="input-field input-field__description-middle"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        why
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                      <span
                        className="input-field--title--counter"
                      >
                        0
                        /
                        500
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <textarea
                        className="input-field--content__text"
                        name="task[why]"
                        onChange={[Function]}
                        pattern=".*"
                        placeholder="Let people know what will be the result of this task being completed (ex: When task is finished, the result will be xxxx)"
                        readOnly={false}
                        required={true}
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldDescriptionMiddle>
              <InputFieldDescriptionMiddle
                eventHandler={[Function]}
                name="task[description]"
                placeholder="Provide a longer description about the task, the type of work involved, and how it will relate to the larger batch"
                required={true}
                symbolLimit={500}
                title="description"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__description-middle"
                  copyOnClick={false}
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgInputRef={null}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  imgRequirements=""
                  imgVerifier={[Function]}
                  max=""
                  min=""
                  name="task[description]"
                  pattern=".*"
                  placeholder="Provide a longer description about the task, the type of work involved, and how it will relate to the larger batch"
                  readOnly={false}
                  recommended={false}
                  required={true}
                  selectEntries={Array []}
                  step=""
                  symbolLimit={500}
                  title="description"
                  type="textarea"
                  value=""
                >
                  <div
                    className="input-field input-field__description-middle"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        description
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                      <span
                        className="input-field--title--counter"
                      >
                        0
                        /
                        500
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <textarea
                        className="input-field--content__text"
                        name="task[description]"
                        onChange={[Function]}
                        pattern=".*"
                        placeholder="Provide a longer description about the task, the type of work involved, and how it will relate to the larger batch"
                        readOnly={false}
                        required={true}
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldDescriptionMiddle>
              <InputFieldDescription
                eventHandler={[Function]}
                name="task[requirements]"
                placeholder="This section is free text that allows for the use of markdown. Create bullets using an asterick and a space before each sentence. Make sure to list a bullet point for each acceptance criteria. These bullet points will be used by reviewers to verify the work."
                required={true}
                symbolLimit={750}
                title="acceptance requirements"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__description"
                  copyOnClick={false}
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgInputRef={null}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  imgRequirements=""
                  imgVerifier={[Function]}
                  max=""
                  min=""
                  name="task[requirements]"
                  pattern=".*"
                  placeholder="This section is free text that allows for the use of markdown. Create bullets using an asterick and a space before each sentence. Make sure to list a bullet point for each acceptance criteria. These bullet points will be used by reviewers to verify the work."
                  readOnly={false}
                  recommended={false}
                  required={true}
                  selectEntries={Array []}
                  step=""
                  symbolLimit={750}
                  title="acceptance requirements"
                  type="textarea"
                  value=""
                >
                  <div
                    className="input-field input-field__description"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        acceptance requirements
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                      <span
                        className="input-field--title--counter"
                      >
                        0
                        /
                        750
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <textarea
                        className="input-field--content__text"
                        name="task[requirements]"
                        onChange={[Function]}
                        pattern=".*"
                        placeholder="This section is free text that allows for the use of markdown. Create bullets using an asterick and a space before each sentence. Make sure to list a bullet point for each acceptance criteria. These bullet points will be used by reviewers to verify the work."
                        readOnly={false}
                        required={true}
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldDescription>
              <InputFieldWhiteDark
                eventHandler={[Function]}
                min="0"
                name="task[amount]"
                placeholder="The total amount of tokens or coins you are paying for this task to be completed"
                required={true}
                step="NaN"
                symbolLimit={0}
                title="award amount (undefined)"
                type="number"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__white__dark"
                  copyOnClick={false}
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgInputRef={null}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  imgRequirements=""
                  imgVerifier={[Function]}
                  max=""
                  min="0"
                  name="task[amount]"
                  pattern=".*"
                  placeholder="The total amount of tokens or coins you are paying for this task to be completed"
                  readOnly={false}
                  recommended={false}
                  required={true}
                  selectEntries={Array []}
                  step="NaN"
                  symbolLimit={0}
                  title="award amount (undefined)"
                  type="number"
                  value=""
                >
                  <div
                    className="input-field input-field__white__dark"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        award amount (undefined)
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <input
                        className="input-field--content__text"
                        max=""
                        min="0"
                        name="task[amount]"
                        onChange={[Function]}
                        onClick={null}
                        placeholder="The total amount of tokens or coins you are paying for this task to be completed"
                        readOnly={false}
                        required={true}
                        step="NaN"
                        type="number"
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldWhiteDark>
              <InputFieldWhiteDark
                eventHandler={[Function]}
                name="task[proof_link]"
                placeholder="URL to a Dropbox or Drive Folder, or a GitHub Repo"
                required={true}
                symbolLimit={150}
                title="where to submit completed work"
                value=""
              >
                <InputField
                  checkboxText=""
                  checked={false}
                  className="input-field__white__dark"
                  copyOnClick={false}
                  disabled={false}
                  errorText=""
                  eventHandler={[Function]}
                  imgInputRef={null}
                  imgPreviewDimensions="40x40"
                  imgPreviewUrl=""
                  imgRequirements=""
                  imgVerifier={[Function]}
                  max=""
                  min=""
                  name="task[proof_link]"
                  pattern=".*"
                  placeholder="URL to a Dropbox or Drive Folder, or a GitHub Repo"
                  readOnly={false}
                  recommended={false}
                  required={true}
                  selectEntries={Array []}
                  step=""
                  symbolLimit={150}
                  title="where to submit completed work"
                  type="text"
                  value=""
                >
                  <div
                    className="input-field input-field__white__dark"
                  >
                    <div
                      className="input-field--title"
                    >
                      <span
                        className="input-field--title--title"
                      >
                        where to submit completed work
                      </span>
                      <span
                        className="input-field--title--required"
                      >
                        required
                      </span>
                      <span
                        className="input-field--title--counter"
                      >
                        0
                        /
                        150
                      </span>
                    </div>
                    <div
                      className="input-field--content"
                    >
                      <input
                        className="input-field--content__text"
                        name="task[proof_link]"
                        onChange={[Function]}
                        onClick={null}
                        pattern=".*"
                        placeholder="URL to a Dropbox or Drive Folder, or a GitHub Repo"
                        readOnly={false}
                        required={true}
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                </InputField>
              </InputFieldWhiteDark>
              <input
                name="authenticity_token"
                readOnly={true}
                type="hidden"
                value="00"
              />
            </form>
          </div>
        </div>
      </div>
      <div
        className="layout--subfooter"
      >
        <div
          className="layout--subfooter--buttons"
        >
          <Button
            className=""
            disabled={false}
            form="task-form--form"
            onClick={[Function]}
            type="submit"
            value="create & close"
          >
            <input
              className="button"
              disabled={false}
              form="task-form--form"
              onClick={[Function]}
              type="submit"
              value="create & close"
            />
          </Button>
          <ButtonBorder
            className=""
            form="task-form--form"
            onClick={[Function]}
            type="submit"
            value="create"
          >
            <Button
              className="button__border"
              disabled={false}
              form="task-form--form"
              onClick={[Function]}
              type="submit"
              value="create"
            >
              <input
                className="button button__border"
                disabled={false}
                form="task-form--form"
                onClick={[Function]}
                type="submit"
                value="create"
              />
            </Button>
          </ButtonBorder>
          <ButtonBorder
            className=""
            onClick={[Function]}
            value="cancel"
          >
            <Button
              className="button__border"
              disabled={false}
              onClick={[Function]}
              type="button"
              value="cancel"
            >
              <input
                className="button button__border"
                disabled={false}
                onClick={[Function]}
                type="button"
                value="cancel"
              />
            </Button>
          </ButtonBorder>
        </div>
      </div>
    </div>
  </Layout>
</TaskForm>
`;
